programs/bpf_loader/src/syscalls.rs

```diff
@@ -2636,6 +2659,8 @@ fn call<'a, 'b: 'a>(
         )
         .map_err(SyscallError::InstructionError)?;
 
+    invoke_context.add_sibling_instruction(instruction);
```

Comment:
Two things:
- Can we use `CompiledInstruction` which has account indices, instead of `Pubkey`s? (better for ABIv2).
- How about referencing the recorded instructions by index / a range?

Reply:
Instructions aren't always recorded, are you thinking we should always turn on full instruction recording?

I'll take a look at using `CompiledInstruction`

Reply:
[This refactoring](https://github.com/solana-labs/solana/pull/22111) should have made it cheap enough to always run it. And this PR here comes close to "always on" as well. So why not have them both use an unified underlying mechanism?

Reply:
Yeah, I'll take a look at that, this implementation is pretty cheap as is, but we can probably do better

Reply:
One thing to keep in mind is that the instruction recorder records instructions before they are invoked, so it includes instructions that haven't been fully processed yet which also results in a different order

Reply:
Recorded inner instructions are defined being invoked, not successfully processed:
https://docs.solana.com/developing/clients/jsonrpc-api#inner-instructions-structure